File: examples/python/fs_ops.py
"""Filesystem operations — demonstrates hostfs mount."""

import os
import tempfile

# Write a file, read it back, stat it.
path = "/mnt/host/hello.txt"
with open(path, "w") as f:
    f.write("Hello from the guest!\n")

with open(path) as f:
    content = f.read()
    print(f"Read back: {content.strip()}")

info = os.stat(path)
print(f"Size: {info.st_size} bytes")

# List the mount directory.
entries = os.listdir("/mnt/host")
print(f"Files in /mnt/host: {entries}")

# Create a subdirectory and a file inside it.
os.makedirs("/mnt/host/subdir", exist_ok=True)
with open("/mnt/host/subdir/nested.txt", "w") as f:
    f.write("nested content\n")

for root, dirs, files in os.walk("/mnt/host"):
    for name in files:
        full = os.path.join(root, name)
        print(f"  {full} ({os.path.getsize(full)} bytes)")

# Write a sentinel that the host can verify.
with open("/mnt/host/sentinel.txt", "w") as f:
    f.write("guest-was-here\n")

# Clean up everything except the sentinel.
os.unlink("/mnt/host/subdir/nested.txt")
os.rmdir("/mnt/host/subdir")
os.unlink(path)
print("Cleanup done.")

File: examples/node/fs_ops.js
// Filesystem operations — demonstrates hostfs mount.

const fs = require('fs');
const path = require('path');

const mountDir = '/mnt/host';
const filePath = path.join(mountDir, 'hello.txt');

// Write and read back.
fs.writeFileSync(filePath, 'Hello from Node.js guest!\n');
const content = fs.readFileSync(filePath, 'utf-8');
console.log(`Read back: ${content.trim()}`);

// Stat.
const stat = fs.statSync(filePath);
console.log(`Size: ${stat.size} bytes`);

// List directory.
const entries = fs.readdirSync(mountDir);
console.log(`Files in ${mountDir}: ${entries.join(', ')}`);

// Create subdirectory.
const subdir = path.join(mountDir, 'subdir');
fs.mkdirSync(subdir, { recursive: true });
fs.writeFileSync(path.join(subdir, 'nested.txt'), 'nested content\n');

// Walk and print.
function walk(dir) {
    for (const entry of fs.readdirSync(dir)) {
        const full = path.join(dir, entry);
        const s = fs.statSync(full);
        if (s.isDirectory()) {
            walk(full);
        } else {
            console.log(`  ${full} (${s.size} bytes)`);
        }
    }
}
walk(mountDir);

// Clean up.
fs.unlinkSync(path.join(subdir, 'nested.txt'));
fs.rmdirSync(subdir);
fs.unlinkSync(filePath);
console.log('Cleanup done.');
